### PR TITLE
Templating: Make theme optional, pass field to fallback caller

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

This PR aligns the code to be consistent with comments and adds a more fully-featured fallback.

**Why do we need this feature?**

To be honest, we don't. This is something I noticed in other work and wanted to see what people thought about it.

**Who is this feature for?**

See above.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/field/displayProcessor.ts [hunk 1]
```diff
@@ -44,7 +45,8 @@ export function getDisplayProcessor(options?: DisplayProcessorOptions): DisplayP
 
   const field = options.field as Field;
   const config = field.config ?? {};
-  const { palette } = options.theme.visualization;
+  const theme = options.theme || createTheme({ colors: { mode: 'dark' } });
```
Comment (gelicia) [inline packages/grafana-data/src/field/displayProcessor.ts:48]: Consistent with comment on line 25

File: public/app/features/templating/dataMacros.test.ts [hunk 2]
```diff
@@ -101,7 +101,7 @@ describe('dataMacros', () => {
 
     const scopedVars = { __dataContext: dataContext };
 
-    expect(_templateSrv.replace('${__data.fields[1]}', scopedVars)).toBe('10000');
+    expect(_templateSrv.replace('${__data.fields[1]}', scopedVars)).toBe('1969-12-31 18:00:10.000');
```
Comment (academo) [inline public/app/features/templating/dataMacros.test.ts:104]: Is this related to the PR? Why does this value change?

Comment (gelicia) [reply public/app/features/templating/dataMacros.test.ts:104]: Field 1 in this test is a date, with the type of `FieldType.date` https://github.com/grafana/grafana/blob/main/public/app/features/templating/dataMacros.test.ts#L26

The values of this field are `5000` and `10000` which are interpreted as timestamps now that the field type is passed in. I'm guessing this timestamp is 10,000 seconds from epoch.

File: packages/grafana-data/src/field/displayProcessor.ts (hunk 1 above)
Comment (ifrost) [reply packages/grafana-data/src/field/displayProcessor.ts:48]: It used to be optional, then it was marked as required in https://github.com/grafana/grafana/pull/33502. There's no description in that PR and I have no context, but I wonder if that was intentional to make it a required param and the comment is wrong 🤔  ?

Comment (dprokop) [reply packages/grafana-data/src/field/displayProcessor.ts:48]: I guess it's because the display processor in it's shape actually has to have a theme provided to get the correct mappings calculations for instance. But with this default fallback it may make sense.

**Review by dprokop — APPROVED** [review]
I think this makes sense. Since the fallback is now part of the implementation, would you mind updating usage like this `theme: theme ?? createTheme(),` so that the new signature is respected?

File: packages/grafana-data/src/field/displayProcessor.ts (hunk 1 above)
Comment (mckn) [reply packages/grafana-data/src/field/displayProcessor.ts:48]: It would be better if we keep it as required so it needs to be passed. We are currently looking into ways to move the "lazy creation" of the theme  context from `@grafana/data` to core when the app starts.

**Review by mckn — CHANGES_REQUESTED** [review]
I'm blocking it for now until we have reached further with the "move theme context" work.

**Comment by gelicia** [comment]
We didnt really need this, so folks can use this work if needed going forward.
